stdlib/src/collections/list.mojo

```diff
@@ -144,16 +171,34 @@ struct List[T: CollectionElement](CollectionElement, Sized, Boolable):
         self.data = unsafe_pointer
         self.size = size
         self.capacity = capacity
+        self._small_buffer = Self._small_buffer_type(uninitialized=True)
+
+    @always_inline
+    fn _sbo_is_in_use(self) -> Bool:
+        @parameter
+        if not Self.sbo_enabled:
+            return False
+        return self.data == self._small_buffer.unsafe_ptr()
 
     fn __moveinit__(inout self, owned existing: Self):
         """Move data of an existing list into a new one.
 
         Args:
             existing: The existing list.
         """
-        self.data = existing.data
         self.size = existing.size
         self.capacity = existing.capacity
+        self._small_buffer = existing._small_buffer
+
+        @parameter
+        if Self.sbo_enabled:
```

Comment:
Nit: SBO is a common optimization that it would be great if we had an internal abstraction that we use to make it easier to plug into existing collection implementations.  For example, we could have a `Layout` class that handles all the SBO storage complexity.

Reply:
Sure, I don't mind making something more generic. Do you have an example in mind of what Mojo collection could benefit from SBO? As far as I know, all other dynamic collections use a `List` behind the scenes, this is the most basic type.